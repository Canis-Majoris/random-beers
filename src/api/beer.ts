import { endpoints } from '@Config';
import { IBeer } from '@Types/beer';

export type RandomBeerResponse = IBeer;

/**
 * Fetches random beer.
 *
 * @return {Promise<IBeer>} beer promise.
 */
export const fetchRandomBeer = async (): Promise<RandomBeerResponse> => {
  const response = await fetch(endpoints.beer.random);
  const json = await response.json();
  return json[0];
};
