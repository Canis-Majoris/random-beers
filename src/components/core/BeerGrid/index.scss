.beer-grid {
  display: flex;
  flex-direction: column;
  align-items: center;

  &-item {
    display: flex;
    flex-direction: row;
    height: $card-height-base;
    column-gap: $space-xs;

    &-image {
      aspect-ratio: 1;

      & > img {
        border-radius: $border-radius-sm;
        width: 100%;
        height: 100%;
        object-fit: contain;
      }
    }

    &-info {
      flex: 1;
      display: flex;
      flex-direction: column;

      h4,
      p {
        margin: 0 0 $space-sm;
      }
    }
  }
}
