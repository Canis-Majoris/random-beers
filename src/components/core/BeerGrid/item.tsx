import { FC } from 'react';
import Card from '@Components/core/Card';
import './index.scss';

export interface BeerGridItemProps {
  name: string;
  image: string;
  tagline?: string;
}

const BeerGridItem: FC<BeerGridItemProps> = ({ image, name, tagline }) => (
  <Card className='beer-grid-item'>
    <div className='beer-grid-item-image'>
      <img src={image} alt={name} />
    </div>
    <div className='beer-grid-item-info'>
      <h4 className='ellipsis'>{name}</h4>
      <p className='ellipsis'>{tagline}</p>
    </div>
  </Card>
);

export default BeerGridItem;
