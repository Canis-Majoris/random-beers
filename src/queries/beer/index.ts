export {
  default as useRandomBeersQuery,
  getUniqueBeersList,
} from './useRandomBeersQuery';
