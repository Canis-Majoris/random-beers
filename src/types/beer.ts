export interface IBeer {
  id: number;
  name: string;
  tagline: string;
  image_url: string;
  first_brewed: string;
}
