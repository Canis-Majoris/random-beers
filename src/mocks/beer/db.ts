import { IBeer } from '@Types/beer';
import { getRandomDate } from '@Utils/date';

// Create mock beer response
export const getBeerMock = (): IBeer => ({
  id: Math.floor(Math.random() * 100000000),
  name: 'Test Beer',
  tagline: 'Beer tagline text',
  image_url:
    'https://images2.minutemediacdn.com/image/upload/c_crop,h_1414,w_2101,x_9,y_0/v1617716394/shape/mentalfloss/643219-gettyimages-1248993201.jpg?itok=NRWUukt9',
  first_brewed: getRandomDate(new Date(1800, 0, 1), new Date()).toString(),
});
