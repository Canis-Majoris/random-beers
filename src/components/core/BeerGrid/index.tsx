import { FC } from 'react';
import { useRandomBeersQuery } from '@Queries/beer';
import { Loading, Alert, Empty, BeerCard } from '@Components/core';
import './index.scss';

export interface BeerGridProps {
  number?: number;
}

const BeerGrid: FC<BeerGridProps> = ({ number = 30 }) => {
  const { data, isLoading, isFetching, error } = useRandomBeersQuery(number);

  return (
    <section className='beer-grid'>
      {(isLoading || isFetching) && <Loading />}
      {!(isLoading || isFetching) && error && (
        <Alert variant='error'>{error}</Alert>
      )}
      {!(isLoading || isFetching) && !error && data?.length === 0 && (
        <Empty>No Beers where retrieved :(</Empty>
      )}
      {!isLoading && (
        <div className='row'>
          {data?.map(({ name, id, image_url, tagline }) => (
            <div key={id} className='col-12 col-desktop-6'>
              <BeerCard name={name} tagline={tagline} image={image_url} />
            </div>
          ))}
        </div>
      )}
    </section>
  );
};

export default BeerGrid;
