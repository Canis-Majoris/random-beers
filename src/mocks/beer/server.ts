import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { endpoints } from '@Config';
import { getBeerMock } from './db';

// for node/test environments
export const server = setupServer(
  rest.get(endpoints.beer.random, (req, res, ctx) => {
    return res(ctx.json([getBeerMock()]));
  })
);
