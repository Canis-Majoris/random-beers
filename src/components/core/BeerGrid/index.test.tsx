import '@testing-library/jest-dom';
import { prettyDOM, render, screen } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { server } from '@Mocks/beer/server';
import { getBeerMock } from '@Mocks/beer/db';
import BeerGrid from '.';

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test('renders BeerGrid component correctly', async () => {
  const queryClient = new QueryClient();
  const { container } = render(
    <QueryClientProvider client={queryClient}>
      <BeerGrid number={50} />
    </QueryClientProvider>
  );
  const beerMock = getBeerMock();
  const beerCards = await screen.findAllByText(beerMock.name, { exact: false });

  expect(beerCards.length).toEqual(50);
  expect(prettyDOM(container)).toMatchSnapshot();
});
