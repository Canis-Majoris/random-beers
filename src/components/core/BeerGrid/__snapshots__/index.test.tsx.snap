// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders BeerGrid component correctly 1`] = `
"[36m<div>[39m
  [36m<section[39m
    [33mclass[39m=[32m\\"beer-grid\\"[39m
  [36m>[39m
    [36m<div[39m
      [33mclass[39m=[32m\\"row\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"col-12 col-desktop-6\\"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"card beer-card\\"[39m
        [36m>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"beer-card-image\\"[39m
          [36m>[39m
            [36m<img[39m
              [33malt[39m=[32m\\"Test Beer\\"[39m
              [33msrc[39m=[32m\\"https://images2.minutemediacdn.com/image/upload/c_crop,h_1414,w_2101,x_9,y_0/v1617716394/shape/mentalfloss/643219-gettyimages-1248993201.jpg?itok=NRWUukt9\\"[39m
            [36m/>[39m
          [36m</div>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"beer-card-info\\"[39m
          [36m>[39m
            [36m<h4[39m
              [33mclass[39m=[32m\\"ellipsis\\"[39m
            [36m>[39m
              [0mTest Beer[0m
            [36m</h4>[39m
            [36m<p[39m
              [33mclass[39m=[32m\\"ellipsis\\"[39m
            [36m>[39m
              [0mBeer tagline text[0m
            [36m</p>[39m
          [36m</div>[39m
        [36m</div>[39m
      [36m</div>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"col-12 col-desktop-6\\"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"card beer-card\\"[39m
        [36m>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"beer-card-image\\"[39m
          [36m>[39m
            [36m<img[39m
              [33malt[39m=[32m\\"Test Beer\\"[39m
              [33msrc[39m=[32m\\"https://images2.minutemediacdn.com/image/upload/c_crop,h_1414,w_2101,x_9,y_0/v1617716394/shape/mentalfloss/643219-gettyimages-1248993201.jpg?itok=NRWUukt9\\"[39m
            [36m/>[39m
          [36m</div>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"beer-card-info\\"[39m
          [36m>[39m
            [36m<h4[39m
              [33mclass[39m=[32m\\"ellipsis\\"[39m
            [36m>[39m
              [0mTest Beer[0m
            [36m</h4>[39m
            [36m<p[39m
              [33mclass[39m=[32m\\"ellipsis\\"[39m
            [36m>[39m
              [0mBeer tagline text[0m
            [36m</p>[39m
          [36m</div>[39m
        [36m</div>[39m
      [36m</div>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"col-12 col-desktop-6\\"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"card beer-card\\"[39m
        [36m>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"beer-card-image\\"[39m
          [36m>[39m
            [36m<img[39m
              [33malt[39m=[32m\\"Test Beer\\"[39m
              [33msrc[39m=[32m\\"https://images2.minutemediacdn.com/image/upload/c_crop,h_1414,w_2101,x_9,y_0/v1617716394/shape/mentalfloss/643219-gettyimages-1248993201.jpg?itok=NRWUukt9\\"[39m
            [36m/>[39m
          [36m</div>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"beer-card-info\\"[39m
          [36m>[39m
            [36m<h4[39m
              [33mclass[39m=[32m\\"ellipsis\\"[39m
            [36m>[39m
              [0mTest Beer[0m
            [36m</h4>[39m
            [36m<p[39m
              [33mclass[39m=[32m\\"ellipsis\\"[39m
            [36m>[39m
              [0mBeer tagline text[0m
            [36m</p>[39m
          [36m</div>[39m
        [36m</div>[39m
      [36m</div>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"col-12 col-desktop-6\\"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"card beer-card\\"[39m
        [36m>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"beer-card-image\\"[39m
          [36m>[39m
            [36m<img[39m
              [33malt[39m=[32m\\"Test Beer\\"[39m
              [33msrc[39m=[32m\\"https://images2.minutemediacdn.com/image/upload/c_crop,h_1414,w_2101,x_9,y_0/v1617716394/shape/mentalfloss/643219-gettyimages-1248993201.jpg?itok=NRWUukt9\\"[39m
            [36m/>[39m
          [36m</div>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"beer-card-info\\"[39m
          [36m>[39m
            [36m<h4[39m
              [33mclass[39m=[32m\\"ellipsis\\"[39m
            [36m>[39m
              [0mTest Beer[0m
            [36m</h4>[39m
            [36m<p[39m
              [33mclass[39m=[32m\\"ellipsis\\"[39m
            [36m>[39m
              [0mBeer tagline text[0m
            [36m</p>[39m
          [36m</div>[39m
        [36m</div>[39m
      [36m</div>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"col-12 col-desktop-6\\"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"card beer-card\\"[39m
        [36m>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"beer-card-image\\"[39m
          [36m>[39m
            [36m<img[39m
              [33malt[39m=[32m\\"Test Beer\\"[39m
              [33msrc[39m=[32m\\"https://images2.minutemediacdn.com/image/upload/c_crop,h_1414,w_2101,x_9,y_0/v1617716394/shape/mentalfloss/643219-gettyimages-1248993201.jpg?itok=NRWUukt9\\"[39m
            [36m/>[39m
          [36m</div>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"beer-card-info\\"[39m
          [36m>[39m
            [36m<h4[39m
              [33mclass[39m=[32m\\"ellipsis\\"[39m
            [36m>[39m
              [0mTest Beer[0m
            [36m</h4>[39m
            [36m<p[39m
              [33mclass[39m=[32m\\"ellipsis\\"[39m
            [36m>[39m
              [0mBeer tagline text[0m
            [36m</p>[39m
          [36m</div>[39m
        [36m</div>[39m
      [36m</div>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"col-12 col-desktop-6\\"[39m
      [36m>[39m
        [36m<div[39m
          [33mclass[39m=[32m\\"card beer-card\\"[39m
        [36m>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"beer-card-image\\"[39m
          [36m>[39m
            [36m<img[39m
              [33malt[39m=[32m\\"Test Beer\\"[39m
              [33msrc[39m=[32m\\"https://images2.minutemediacdn.com/image/upload/c_crop,h_1414,w_2101,x_9,y_0/v1617716394/shape/mentalfloss/643219-gettyimages-1248993201.jpg?itok=NRWUukt9\\"[39m
            [36m/>[39m
          [36m</div>[39m
          [36m<div[39m
            [33mclass[39m=[32m\\"beer-card-info\\"[39m
          [36m>[39m
            [36m<h4[39m
              [33mclass[39m=[32m\\"ellipsis\\"[39m
            [36m>[39m
              [0mTest Beer[0m
            [36m</h4>[39m
            [36m<p[39m
              [33mclass[39m=[32m\\"ellipsis\\"[39m
            [36m>[39m
              [0mBeer tagline text[0m
            [36m</p>[39m
          [36m</div>[39m
        [36m</div>[39m
      [36m</div>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"col-12 col-desktop-6\\"[39m
      [36m>..."
`;
